docs/internals/RFC-feature-coverage.md

```diff
@@ -0,0 +1,168 @@
+## TL;DR.
+
+We'll implement a mechanism that will put metadata on each test class about the coverage/completness of the test. Two questions : 
+
+* There will be three level, `basic`, `good`, `complete`. It's a compromise, do you think it's too many? not enough?
+* There two proposition of implementations, one , one using class inheritance : 
+
+
+```python
+@coverage.good
+class Test_Feature():
+    pass
+```
+
+Or
+
+```python
+class Test_Feature(BaseTestCase.GoodCoverage):
+    pass
+```
+
+There are [pro and con](#Implementations possibility), which one do you prefer ? 
+
+## Currently
+
+For a given feature, a test can have two informal coverage status:
+
+* The test is written. It provides a certain level of garantee for the feature to be bug-free
+* The test is not written
+
+The first status is obvious to observe (the test class exists), **but** there is no information about how far the test cover the feature. From bottom to top, it can be : 
+
+* just a declarative test (nothing is actually tested)
+* indirect testing, it guess that the feature is working becase another one is working
+* a simple smoke test
+* a quite complete test
+* a rock solid test, implying a very strong confidence on the feature
+
+The second state is more ambigious. Often, the test classe is missing, but we don't know if it's because the feature is not testable or if we can one day implement it.
+
+This document will propose a mechanism to add metadata on test classes that allow to describe precisely the test coverage of each feature
+
+## Possible test status
+
+### Not testable
+
+System test outputs are exported and used in metabase or google spreadsheets. As some feature are not testable, those UI needs to mix declartive data, and system test output. As an declarative test class is not harder than 2 code line, having this status will help us to have a complete automatic UI for feature status.
+
+Furthermore, a feature is not testable, until we find a way to test them. With this status, we will save times and effort if it happens, as we won't have anything to change in system test output process.
+
+### Not implemented
+
+Obviously, it means that the test is not yet implemented
+
+### Basic test
+
+The test is the most simple test we can imagine for a feature. It ensure that the feature is working, but does not cover any subtility or edge cases
+
+### Good test
+
+The test fully covers the core functionnality of the feature. The test will catch all bug that can happen, except for edge cases and extreme scenarios
+
+### Complete test
+
+The test fully cover the functionnality, and missing a bug is statistically negligible => If a bug is missed, we should be included in the post mortem and review our understanding of the feature
```

Comment:
Very interesting subject :) If by "complete" we mean "perfect", you're perfectly true. A test will never be perfect.

What I mean here is more "We are very confident that a bug cannot happen on this feature if the test passes." Maybe `complete` is not the best word ? (ping @lloeki who loves finding good naming ;) )

Reply:
I was just thinking about that @simon-id, while trying to understand the exact meaning behind these names and descriptions! ;)

Indeed I think naming this way is definitively very fuzzy, which makes it open to interpretation on both ends (test writer and test result reader). What threshold makes it not "basic" anymore? What is "good": good enough or close to complete? Can a test be reasonably marked as "complete", what if we find a new bug or class of bugs, should "complete" tests that passed with the bug stop being "complete" until the corresponding case is implemented?

@cbeauchesne I think you introduced very interesting, unambiguous descriptions right above, in ways that are clear-cut and barely leave room for interpretation, thus carrying good semantic value:

- `indirect`: the test works by inferring a result from state/globals/behaviour that is not related to the feature itself, i.e using inductive logic. consequence: unrelated moving parts may break the test, but having it pass is a good signal that things are probably working as expected
- `smoke`: the test is extremely minimal and essentially checks whether hitting the feature does not produce breakage: it doesn't test that the feature works, it tests that the feature isn't 
- `minimal`: smoke test + tests that the feature actually works, in the smallest possible sense, e.g happiest path, smallest non-thorough value check. rule of thumb: if you remove anything it doesn't test "feature is working" anymore and thus reverts back to be a smoke test; hence "minimal".
- `happy`: Attempts to test multiple happy paths beyond minimal
- `unhappy`: Attempts to test one or more unhappy paths
- `corners`: Attempts to test identified corner cases, this qualifier does not reflect thoroughness of testing each possible corner, it reflects the difference between testing the most common happy+unhappy paths vs going beyond, with specific, trickier cases.
- `thorough`: Happy and unhappy paths are considered broadly tested. Might be an attribute instead of a decorator, e.g `happy(thorough: true)`, `corners(thorough: true)`...

These could be combinable, e.g:
- minimal+happy+unhappy: both happy and unhappy are minimally tested in a non-inferred way
- happy(thorough)+corners: many positive tests, no negative tests, a few corner cases
- indirect+smoke: smoke tested via inference

I feel this way we can eyeball testing coverage with confidence, possibly even compute a synthetic per test+total score from it.

Reply:
Only my opinions:
Loic has something, maybe too complex (I fear that we did not fill well the metadata in the long run, if there are too many of them). On the other hand, I think Loic's proposal is relevant. We should think about the complexity of usage vs quality of insights.
If we are ready to pay this complexity, I would support Loic model. Every value could be a decorator.

@lloeki I learn something today (I was not aware of this terminology happy/unhappy path). Thanks for sharing.

Reply:
Actually, for functional coverage, I think that we'll never be able to get rid of implicit thresholds : we'll never have 100% complete RFC describing every possible use cases, and giving for each of them a notation, so we'll always need to delegate to the test which use case of a feature is a corner cas, or a happy path, etc. And even with this details, you still have an implicit threshold : *are considered broadly tested*.

I totally agree that those thresholds are the pitfall of this feature. But as we cannot eliminate them, my strategy is to limit them to a level of details compatible with how often developers write tests.

One key thing to keep in mind is that they write or modify system tests not so often. Maybe once per month for AppSec engineers. And giving this notation will be done only for those who wrote the test, not when you only enable it.

So having too many detail level will cause a lot of friction, and at the end will give exactly the opposite expected result : no notation or wrongly interpreted notation.

If one day we have a dedicated team to write test, I will agree that more details will be far better, as this team will use them daily. But for now, I think that keeping it simple is very important. I even hesitated with only two levels, using your terms, would be `minimal` and `not-minimal` 😄